src/object.d

```diff
@@ -4912,17 +4911,15 @@ private void onArrayCastError()(string fromType, size_t fromSize, string toType,
         , unsignedToTempString(toSize)
     ];
 
-    // convert discontiguous `msgComponents` to contiguous string on the stack
-    size_t length = 0;
-    foreach (m ; msgComponents)
-        length += m.length;
-
-    auto msg = (cast(char*)alloca(length))[0 .. length];
+    // `alloca` would be preferred here, but it doesn't work in -betterC
+    // See https://issues.dlang.org/show_bug.cgi?id=19159
+    char[1024] msg = void;
```

Comment:
This looks like a classic buffer overrun opportunity. With templates type names in D can easily get pretty long. Instead of trying to find a one size fits all number, summing the string length as before + using dynamic allocation would be more appropriate.